packages/labs/src/controllers/context.ts

```diff
@@ -0,0 +1,324 @@
+import {UpdatingController} from './updating-controller.js';
+export * from './updating-controller.js';
+import {notEqual, UpdatingElement} from 'updating-element';
+import {
+  AttributePart,
+  ATTRIBUTE_PART,
+  directive,
+  Directive,
+  nothing,
+  Part,
+  PROPERTY_PART,
+  DirectiveClass,
+  DirectiveParameters,
+} from 'lit-html';
+
+// TODO(sorvell): lit-html should export this?
+declare type DirectiveResult<C extends DirectiveClass = DirectiveClass> = {
+  _$litDirective$: C;
+  values: DirectiveParameters<C>;
+};
+
+const connectContextEvent = 'connect-context';
+
+const providerMap: WeakMap<Node, Provider[]> = new WeakMap();
+
+export type ConsumerEvent = CustomEvent<Consumer[]>;
+
+export const EVENT_STRATEGY = 1;
+export const ASCEND_STRATEGY = 2;
+
+const DEFAULT_STRATEGY = EVENT_STRATEGY;
+
+// Note, this currently wouldn't strictly have to be an UpdatingController.
+export class Provider extends UpdatingController {
```

Comment:
Can you explain why?

Reply:
I think the contract between provider and consumer should be narrow so that there can be other implementations that don't have to implement all of UpdatingController. If someone gets a reference to a Provider, should they be able to call the controller's lifecycle callbacks?